Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean

```diff
@@ -0,0 +1,199 @@
+
+/-- Given a presheaf of modules `M`, this is a constructor for the type `M.Elements`. -/
+abbrev elementsMk {C : Type u₁} [Category.{v₁} C] {R : Cᵒᵖ ⥤ RingCat.{u}}
+    (M : PresheafOfModules.{v} R) (X : Cᵒᵖ) (x : M.obj X) : M.Elements :=
+  Functor.elementsMk _ X x
+
+namespace Elements
+
+variable {C : Type u} [Category.{v} C] {R : Cᵒᵖ ⥤ RingCat.{v}} {M : PresheafOfModules.{v} R}
+
+/-- Given an element `m : M.Elements` of a presheaf of modules `M`, this is
+free presheaf of modules on the Yoneda presheaf of types corresponding to the
+underlying object of `m`. -/
+noncomputable abbrev freeYoneda (m : M.Elements) :
+    PresheafOfModules.{v} R := (free R).obj (yoneda.obj m.1.unop)
+
+/-- Given an element `m : M.Elements` of a presheaf of modules `M`, this is
+the canonical morphism `m.freeYoneda ⟶ M`. -/
+noncomputable abbrev fromFreeYoneda (m : M.Elements) :
+    m.freeYoneda ⟶ M :=
+  freeYonedaEquiv.symm m.2
+
+lemma fromFreeYoneda_app_apply (m : M.Elements) :
+    m.fromFreeYoneda.app m.1 (ModuleCat.freeMk (𝟙 _)) = m.2 := by
+  apply freeYonedaEquiv_symm_app
+
+end Elements
+
+section
+
+variable {C : Type u} [SmallCategory.{u} C] {R : Cᵒᵖ ⥤ RingCat.{u}} (M : PresheafOfModules.{u} R)
+
+/-- Given a presheaf of modules `M`, this is the coproduct of
+all free Yoneda presheaves `m.freeYoneda` for all `m : M.Elements`. -/
+noncomputable abbrev freeYonedaCoproduct : PresheafOfModules.{u} R :=
+  ∐ (Elements.freeYoneda (M := M))
+
+/-- Given an element `m : M.Elements` of a presheaf of modules `M`, this is the
+canonical inclusion `e.freeYoneda ⟶ M.freeYonedaCoproduct`. -/
+noncomputable abbrev ιFreeYonedaCoproduct (m : M.Elements) :
+    m.freeYoneda ⟶ M.freeYonedaCoproduct :=
+  Sigma.ι _ m
+
+/-- Given a presheaf of modules `M`, this is the
+canonical morphism `M.freeYonedaCoproduct ⟶ M`. -/
+noncomputable def fromFreeYonedaCoproduct :
+    M.freeYonedaCoproduct ⟶ M :=
+  Sigma.desc Elements.fromFreeYoneda
+
+/-- Given an element `m` of a presheaf of modules `M`, this is the associated
+canonical section of the presheaf `M.freeYonedaCoproduct` over the object `m.1`. -/
+noncomputable def freeYonedaCoproductMk (m : M.Elements) :
+    M.freeYonedaCoproduct.obj m.1 :=
+  (M.ιFreeYonedaCoproduct m).app _ (ModuleCat.freeMk (𝟙 _))
+
+@[reassoc (attr := simp)]
+lemma ι_fromFreeYonedaCoproduct (e : M.Elements) :
+    M.ιFreeYonedaCoproduct e ≫ M.fromFreeYonedaCoproduct = e.fromFreeYoneda := by
+  apply Sigma.ι_desc
+
+lemma ι_fromFreeYonedaCoproduct_apply (e : M.Elements) (X : Cᵒᵖ) (x : e.freeYoneda.obj X) :
+    M.fromFreeYonedaCoproduct.app X ((M.ιFreeYonedaCoproduct e).app X x) =
+      e.fromFreeYoneda.app X x :=
+  congr_fun ((evaluation R X ⋙ forget _).congr_map (M.ι_fromFreeYonedaCoproduct e)) x
+
+@[simp]
+lemma fromFreeYonedaCoproduct_app_mk (e : M.Elements) :
+    M.fromFreeYonedaCoproduct.app _ (M.freeYonedaCoproductMk e) = e.2 := by
+  dsimp [freeYonedaCoproductMk]
+  erw [M.ι_fromFreeYonedaCoproduct_apply e]
```

Comment:
Do you know why `  rw [M.ι_fromFreeYonedaCoproduct_apply e e.fst (ModuleCat.freeMk (𝟙 (Opposite.unop e.fst)))]` doesn't work?

Reply:
This is basically the same reason why ` ι_fromFreeYonedaCoproduct_apply` cannot be a simp lemma, as, according to the linter, it is not in simpNF. If we apply `dsimp` to the statement, for the "human" eye, nothing changes, but there are hidden changes behind a `DFunLike.coe`: the definition of some types can be unfolded by dsimp. See also https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/attribute.20simp_NF_my_lemma.3F/near/443101462